laghos_solver.cpp

```diff
@@ -563,25 +563,37 @@ void LagrangianHydroOperator::ComputeDensity(ParGridFunction &rho) const
    }
 }
 
+
 double LagrangianHydroOperator::InternalEnergy(const ParGridFunction &gf) const
 {
    double glob_ie = 0.0;
-   // This should be turned into a kernel so that it could be displayed in pa
-   if (!p_assembly)
-   {
-      Vector one(l2dofs_cnt), loc_e(l2dofs_cnt);
-      one = 1.0;
-      Array<int> l2dofs;
-      double loc_ie = 0.0;
-      for (int e = 0; e < NE; e++)
-      {
-         L2.GetElementDofs(e, l2dofs);
-         gf.GetSubVector(l2dofs, loc_e);
-         loc_ie += Me(e).InnerProduct(loc_e, one);
-      }
-      MPI_Comm comm = H1.GetParMesh()->GetComm();
-      MPI_Allreduce(&loc_ie, &glob_ie, 1, MPI_DOUBLE, MPI_SUM, comm);
-   }
+   
+    // get the restriction and interpolator objects 
+    const QuadratureInterpolator* l2_interpolator = L2.GetQuadratureInterpolator(ir);
+    l2_interpolator->SetOutputLayout(QVectorLayout::byVDIM);
+    auto L2r = L2.GetElementRestriction(ElementDofOrdering::LEXICOGRAPHIC);
+    const int NQ = ir.GetNPoints(); 
+   Vector e_vector(NE*NQ), eintQ(NE*NQ);
+
+    // Get internal energy at the quadrature points
```

Comment:
Thanks for the addition @varchasgopalaswamy.
Once the `cpu` code is done, we should turn the for-loops into kernels that could run on the `mfem::Device`s.
Let me know if you need extra information for that.

Reply:
So I took a look at how the MFEM_FORALL loops work, and I think I have it working - except that I am not sure how to do reductions inside an MFEM_FORALL. So if I have something like 

```
double integral = 0; 
MFEM_FORALL_2D(e, NE, Q1D, Q1D, 1,
      {
         MFEM_FOREACH_THREAD(qy,y,Q1D)
         {
            MFEM_FOREACH_THREAD(qx,x,Q1D)
            {
               const int q = qx + qy * Q1D;
               double vmag = 0; 
               for(int k = 0; k < VDIM; k++){
                  vmag += pow(f_vals(k,q,e),ln_norm);
               }
               integral += mass_vals(q,e)*vmag;
            }
         }
      });

```

I get the compiler error 
```laghos_solver.cpp:616:25: error: assignment of read-only variable ‘integral’
         integral += mass_vals(q,e)*vmag;
         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
```

which I suppose makes sense since `integral` is defined outside the kernel. Are there any flags I should pass when instantiating `double integral = 0;` or when creating the `MFEM_FORALL` loop?

Reply:
You can look at what is done in [Rho0DetJ0Vol](https://github.com/CEED/Laghos/blob/16fa28294908fd15dda8241d6d6c75ba50a12b50/laghos_solver.cpp#L83) to see how the reduction can be done.

Before entering the `MFEM_FORALL` you must also take care of the pointers by using the appropriate `Read`/`ReadWrite` and `Reshape` if needed: this will help with the compiler error you show.

Reply:
Oh, I think I see. The reduction needs to be done like `volume = vol * one;`?

Reply:
Right - it's a possible way to do, not the only one.